pkg/controllers/externaloidc/externaloidc_controller.go

```diff
@@ -0,0 +1,363 @@
+			if err != nil {
+				errs = append(errs, err)
+				issuerURLValid = false
+			} else {
+				if issuerURL.Scheme != "https" {
+					errs = append(errs, fmt.Errorf("issuer URL must use HTTPS"))
+					issuerURLValid = false
+				}
+				if issuerURL.User != nil {
+					errs = append(errs, fmt.Errorf("URL must not contain a username or password"))
+					issuerURLValid = false
+				}
+				if len(issuerURL.RawQuery) > 0 {
+					errs = append(errs, fmt.Errorf("URL must not contain a query"))
+					issuerURLValid = false
+				}
+				if len(issuerURL.Fragment) > 0 {
+					errs = append(errs, fmt.Errorf("URL must not contain a fragment"))
+					issuerURLValid = false
+				}
+			}
+		}
+
+		if len(jwt.Issuer.Audiences) == 0 {
+			errs = append(errs, fmt.Errorf("at least one audience must be defined"))
+		}
+		seenAudiences := sets.NewString()
+		for i, aud := range jwt.Issuer.Audiences {
+			if len(aud) == 0 {
+				errs = append(errs, fmt.Errorf("audience must not be empty (at index %d)", i))
+			} else if seenAudiences.Has(aud) {
+				errs = append(errs, fmt.Errorf("duplicate audience: %s", aud))
+			}
+
+			seenAudiences.Insert(aud)
+		}
+
+		if len(jwt.Issuer.CertificateAuthority) > 0 {
+			caCertPool, err := cert.NewPoolFromBytes([]byte(jwt.Issuer.CertificateAuthority))
+			if err != nil {
+				errs = append(errs, fmt.Errorf("issuer CA is invalid: %v", err))
+			}
+
+			if issuerURL != nil && issuerURLValid {
+				if err := validateCACert(*issuerURL, caCertPool); err != nil {
+					errs = append(errs, fmt.Errorf("could not validate IDP URL with specified CA cert: %v", err))
+				}
+			}
+		}
+
+		seenClaims := sets.NewString()
+		for i, rule := range jwt.ClaimValidationRules {
+			if len(rule.Claim) == 0 {
+				errs = append(errs, fmt.Errorf("claim must not be empty for claim validation rule at index %d", i))
+			} else if seenClaims.Has(rule.Claim) {
+				errs = append(errs, fmt.Errorf("duplicate claim validation rule: %s", rule.Claim))
+			}
+
+			seenClaims.Insert(rule.Claim)
+		}
+
+		if len(jwt.ClaimMappings.Username.Claim) == 0 {
+			errs = append(errs, fmt.Errorf("username claim must not be empty"))
+		} else if jwt.ClaimMappings.Username.Prefix == nil {
+			errs = append(errs, fmt.Errorf("username prefix must not be nil when claim is set"))
+		}
+
+		if len(jwt.ClaimMappings.Groups.Claim) > 0 && jwt.ClaimMappings.Groups.Prefix == nil {
+			errs = append(errs, fmt.Errorf("group prefix must not be nil when claim is set"))
+		}
+	}
+
+	return
+}
+
+// validateCACert makes a request to the provider's well-known endpoint using the
+// specified CA cert pool to validate that the certs in the pool match the host
+func validateCACert(hostURL url.URL, caCertPool *x509.CertPool) error {
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{RootCAs: caCertPool},
+		},
+		Timeout: 5 * time.Second,
+	}
+
+	wellKnown := strings.TrimSuffix(hostURL.String(), "/") + "/.well-known/openid-configuration"
```

Comment:
This path is a `const` defined by the OIDC spec: `"/.well-known/openid-configuration"`?

Reply:
That's right, this is defined here: https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig

This is also what the `kube-apiserver` is using when creating an instance of the authenticator. SeeL
- https://github.com/openshift/kubernetes-apiserver/blob/7742d76eabcb5f3d49e995b91b499b478d4f2f2e/plugin/pkg/authenticator/token/oidc/oidc.go#L278
- https://github.com/coreos/go-oidc/blob/v2.2.1/oidc.go#L114